src/rules/noUnnecessaryParensRule.ts

```diff
@@ -0,0 +1,190 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { isLiteralExpression, isParenthesizedExpression } from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+interface Options {
+    withChild?: string[];
+    asChildOf?: string[];
+    default?: boolean;
+}
+
+const syntaxKindMapping = ts.SyntaxKind as {} as { [k: string]: ts.SyntaxKind };
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-unnecessary-parens",
+        description: Lint.Utils.dedent`
+            Warns when parentheses are used that are unnecessary`,
+        options: {
+            type: "object",
+            properties: {
+                withChild: {
+                    type: "list",
+                    listType: "string",
+                },
+                asChildOf: {
+                    type: "list",
+                    listType: "string",
+                },
+                default: { type: "boolean" },
+            },
+            additionalProperties: false,
+        },
+        optionsDescription: Lint.Utils.dedent`
+            withChild: A list of token kinds around which to flag parentheses.
+                For example, \`{"withChild": ["Identifier"]}\` would flag
+                \`(foo)\` as having unnecessary parentheses around it.
+
+            asChildOf: A list of token kinds and properties on those tokens
+                such that if the parenthesized expression is the appropriate
+                child of a token of that kind, it will be flagged. For example,
+                \`{"asChildOf": ["VariableDeclaration.initializer"]}\` would
+                flag the parentheses in \`let x = (1 + 2)\`, regardless of the
+                kind of the parenthesized expression.
+
+            default: Whether to default the set of bans to a set of hopefully
+                uncontroversial bans picked by tslint.
+            `,
+        optionExamples: [
+            [{
+                withChild: [
+                    "Identifier",
+                    "LiteralExpression",
+                ],
+                asChildOf: [
+                    "VariableDeclaration.initializer",
+                    "ParenthesizedExpression.expression",
+                    "CallExpression.arguments",
+                    "ExpressionStatement.expression",
```

Comment:
All of these options require knowledge of typescript's AST. I don't think that's practicle for most TSLint users.

Maybe the rule should just find *all* unnecessary parens by default and provide options to disable some of them. Something like `"allow-typeof"` to allow `(typeof foo)==="string"` for example. These can be added on demand.

Also with these options, there's currently no way to ban parens on `"*.type"`, e.g. `let foo: (string)` and `foo(param: (1)) {}`

Reply:
There's currently no way to ban `(1 + 2) + 3` without banning `(1 + 2) * 3`, too.

Reply:
> Maybe the rule should just find all unnecessary parens

So, I didn't want to get into litigating what's "necessary". I don't want to tell people to not do `(true && false) || true` or something, when I think that's probably preferable to leaving out the parentheses since it's hard to remember boolean prescience.

Instead, I want to ban them where they're *clearly* unnecessary, and leave the edge cases for human reviewers (or the judgement of the author).

I feel like having a list of allowances would get out of hand.

If you think this approach is fundamentally unsuitable, that's fine, just let me know. I'll just use it on my project.

> Also with these options, there's currently no way to ban parens on "*.type", e.g. let foo: (string) and foo(param: (1)) {}
 
Why not? I added some examples and tests.

Reply:
Maybe the documentation could point to http://astexplorer.net. That's how I find these AST names.

Reply:
Adding a link to astexplorer.net sounds like a good idea.